sw/host/opentitanlib/src/console/mod.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Result};
+use regex::{Captures, Regex};
+use serde::{Deserialize, Serialize};
+use std::fs::File;
+use std::io::{Read, Write};
+use std::os::unix::io::AsRawFd;
+use std::time::{Duration, Instant, SystemTime};
+use thiserror::Error;
+
+use crate::impl_serializable_error;
+use crate::util::file;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ExitStatus {
+    None,
+    CtrlC,
+    Timeout,
+    ExitSuccess,
+    ExitFailure,
+}
+
+/// Errors related to the console interface.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum ConsoleError {
+    #[error("{0}")]
+    GenericError(String),
+}
+impl_serializable_error!(ConsoleError);
+
+pub trait ConsoleDevice {
+    /// Reads data from the UART to print to the console (used when this UART is the console device).
+    fn console_read(&self, buf: &mut [u8], timeout: Duration) -> Result<usize>;
+
+    /// Writes console input data to the UART (used when this UART is the console device).
+    fn console_write(&self, buf: &[u8]) -> Result<()>;
+}
```

Comment:
Since this is a "base" trait that `Uart` and the spi `Target` traits depend on, this belongs in src/io/ somewhere.

I'd also suggest creating a default implementation in the trait:
```
pub trait ConsoleDevice {
  fn console_read(...) ->Result<usize> {
    Err(UnsupportedError.into())
  }
}
```

That way, you can provide an empty impl for all the versions of spi `Target` that don't support it:
```
impl ConsoleDevice for UltradebugSpiTarget {}
```

Reply:
Moved to `src/io/console.rs` and added default impl, good idea!